### PR TITLE
chore: optimize multi architecture build

### PR DESCRIPTION
**Pull request by prashant-shahi**
This PR optimizes multi architecture build flow by building artifacts outside of container. 
Also, updated the depended workflows along with including intermediate `.build` folder in .gitignore file.

Additionally, sets `/logsmigrations` as default log migration folder from go constant.

Signed-off-by: Prashant Shahi <prashant@signoz.io>

### CODE REVIEW

**Comment by prashant-shahi** [comment]
> Please add some description to give an idea of what has changed

PR description updated

File: exporter/clickhouselogsexporter/factory.go [hunk 1]
```diff
@@ -33,7 +33,7 @@ const (
 	archiveNamespace = "clickhouselogs-archive"
 	databaseName     = "signoz_logs"
 	tableName        = "logs"
-	migrationsFolder = "./migrations"
+	migrationsFolder = "/logsmigrations"
```
Comment (srikanthccv) [inline exporter/clickhouselogsexporter/factory.go:36]: Running binary directly instead of docker image will not work this change?

File: cmd/signozcollector/Dockerfile [hunk 2]
```diff
@@ -1,34 +1,25 @@
...
-ARG USER_UID=10001
-
-USER ${USER_UID}
-
-COPY --from=build /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
-COPY --from=build /out/signoz-collector /
-COPY --from=build /src/exporter/clickhousetracesexporter/migrations /migrations
-COPY --from=build /src/exporter/clickhouselogsexporter/migrations /logsmigrations
-
-ENV LOG_MIGRATIONS_FOLDER="/logsmigrations"
```
Comment (srikanthccv) [inline cmd/signozcollector/Dockerfile:30]: You probably have to revert this change. Try running the binary (important because building new image each time is time consuming) and check if the migrations run for logs.

File: cmd/signozcollector/Dockerfile [hunk 3]
```diff
@@ -1,34 +1,25 @@
...
+# expose OTLP ports for the collector
+EXPOSE 4317
+EXPOSE 4318
 
-# Final stage
-FROM alpine:3.15
-
-ARG USER_UID=10001
-
-USER ${USER_UID}
```
Comment (srikanthccv) [inline cmd/signozcollector/Dockerfile:23]: I forgot why this existed (I think something related ClickHouse), how does this removal of this affect?

File: Makefile [hunk 4]
```diff
@@ -31,7 +34,18 @@ test:
 
 .PHONY: build
 build:
-	GOOS=$(GOOS) GOARCH=$(GOARCH) go build -o signoz-collector ./cmd/signozcollector
+	CC=${GOCC} OS111MODULE=on CGO_ENABLED=1 GOOS=${GOOS} GOARCH=${GOARCH} go build -o .build/${GOOS}-${GOARCH}/signoz-collector -ldflags "-linkmode external -extldflags '-static' -s -w ${LD_FLAGS}" ./cmd/signozcollector
```
Comment (srikanthccv) [inline Makefile:37]: `-tags timetzdata` is missing

File: exporter/clickhouselogsexporter/factory.go (hunk 1 above)
Comment (prashant-shahi) [reply exporter/clickhouselogsexporter/factory.go:36]: In case of running directly, one can set the environment variable:

`export LOG_MIGRATIONS_FOLDER="./logsmigrations"`

For most case that is running in Docker/K8s, this newly introduced changes should work fine.
Also, keeping it consistent with the traces migration constant.

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (prashant-shahi) [reply Makefile:37]: Removing it as redundant. Since, that value should be set as `on` by default i.e. use Go modules.

File: exporter/clickhouselogsexporter/factory.go (hunk 1 above)
Comment (srikanthccv) [reply exporter/clickhouselogsexporter/factory.go:36]: It works fine in docker/k8s, but that isn't my original concern. Generally, programs have defaults, and env/cli args are added to support the customisation. The stand-alone binary should run with defaults without any custom configuration, but this change doesn't make it work.

File: cmd/signozcollector/Dockerfile (hunk 3 above)
Comment (prashant-shahi) [reply cmd/signozcollector/Dockerfile:23]: This is not needed now. This made sense when we were using `scratch` base image, but with `alpine` it is not needed.

File: Makefile (hunk 4 above)
Comment (prashant-shahi) [reply Makefile:37]: nice catch

File: exporter/clickhouselogsexporter/factory.go (hunk 1 above)
Comment (prashant-shahi) [reply exporter/clickhouselogsexporter/factory.go:36]: yeah, true. It's been set to `/migrations` as default for traces.
That should be updated.

File: Makefile (hunk 4 above)
[the remark this answers is not in the document]
Comment (prashant-shahi) [reply Makefile]: updated

File: Makefile [hunk 5]
```diff
@@ -31,7 +34,18 @@ test:
...
+
+.PHONY: amd64
+amd64:
+	make GOARCH=amd64 build
+
+.PHONY: arm64
+arm64:
+	make GOCC=aarch64-linux-gnu-gcc GOARCH=arm64 build
```
[the remark this answers is not in the document]
Comment (prashant-shahi) [reply Makefile]: OR support? you mean across amd/arm? It works on both of those. However, the GOOS env would be set to `linux` for both while building.

File: cmd/signozcollector/Dockerfile [hunk 6]
```diff
@@ -1,34 +1,27 @@
...
 
-ARG TARGETPLATFORM
+# define arguments and default values
+ARG TARGETOS=linux
+ARG TARGETARCH=amd64
 
-RUN export GOOS=$(echo ${TARGETPLATFORM} | cut -d / -f1) && \
-    export GOARCH=$(echo ${TARGETPLATFORM} | cut -d / -f2)
+# copy the binaries from the multi-stage build
+COPY .build/${TARGETOS}-${TARGETARCH}/signoz-collector /signoz-collector
```
Comment (srikanthccv) [inline cmd/signozcollector/Dockerfile:12]: Don't you have to set `TARGETOS` and `TARGETARCH` while building docker images?

Comment (prashant-shahi) [reply cmd/signozcollector/Dockerfile:12]: We do not have to set them. It is automatically set globally by buildkit.

Reference: https://docs.docker.com/engine/reference/builder/#automatic-platform-args-in-the-global-scope

**Comment by prashant-shahi** [comment]
moving the PR branch from fork to main repo and creating another PR
